### ISSUE TITLE
std::regex throws exception in windows 

### ISSUE THREAD

**Issue by ThiagoIze** [OpenImageIO/oiio#2169]
More fun in Windows land: We took a multichannel exr file and generated a tx file (.tiff under the hood), so that later when we operate on the .tx file, for instance by using oiiotool --info, we get a crash. In our application we catch exceptions (side note, should oiiotool catch exceptions too?), so instead of crashing we find this exception:
"regex_error(error_stack): There was insufficient memory to determine whether the regular expression
 could match the specified character sequence."

Debugging points to tiffinput.cpp with the exception being thrown from the `regex_replace` call in:
```c++
    found = desc.rfind("oiio:AverageColor=");
    if (found != std::string::npos) {
        size_t begin  = desc.find_first_of('=', found) + 1;
        size_t end    = std::min(desc.find_first_of(' ', begin), desc.size());
        string_view s = string_view(desc.data() + begin, end - begin);
        m_spec.attribute("oiio:AverageColor", s);
        const std::string average_pattern
            = std::string("oiio:AverageColor=(\\[?") + fp_number_pattern
              + ",?)+\\]?[ ]*";
        printf("desc=%s\n", desc.c_str());
        desc        = regex_replace(desc, regex(average_pattern), "");
        updatedDesc = true;
    }
```

This seems like it could be related to https://stackoverflow.com/questions/42909618/a-simple-c11-regex-that-throws-regex-error-while-searching , but for some reason setting a `#define _REGEX_MAX_STACK_COUNT 0` at the very top of tiffinput.cpp doesn't help even though looking at the regex source code makes it look like it should. In fact, doing that in the simpler patch below also doesn't help. Perhaps something sneaky is going on and this isn't getting used by the windows regex header? It'd be nice to figure this out, but I gave up and went with the below fix.

Since we still have the boost code, the following lets us build with boost whenever this is an issue:
```diff
diff --git a/src/cmake/compiler.cmake b/src/cmake/compiler.cmake
index 37a87776..47e2e233 100644
--- a/src/cmake/compiler.cmake
+++ b/src/cmake/compiler.cmake
@@ -267,6 +267,16 @@ set (CMAKE_REQUIRED_DEFINITIONS ${CSTD_FLAGS})
 check_cxx_source_runs("
       #include <regex>
       int main() {
+
+          // This throws an exception on windows, at least with VS2015, when using std_regex
+          \"desc=oiio:SHA-1=879F1A17ACE62D0B3D6CCE09B6F2BE3ADD6A48B2 oiio:AverageColor=0.532385,0.49592,0.494299,0,1.16619,0.946394,0.784551,0,0.6515
71,0.649502,0.65518,0,0.489643,0.471951,0.471693,0,0.599104,0.547551,0.488581,0,0.796633,0.890403,0.84728,0,0.857453,0.874942,0.906667,0,1.26352,1.247
64,1.2966,0,1.86804,1.62555,1.16578,0,1.88414,1.9104,2.07001,0\";
+          static const char* fp_number_pattern
+              = \"([+-]?((?:(?:[[:digit:]]*\\.)?[[:digit:]]+(?:[eE][+-]?[[:digit:]]+)?)))\";
+          const std::string average_pattern
+              = std::string(\"oiio:AverageColor=(\\[?\") + fp_number_pattern
+              + \",?)+\\]?[ ]*\";
+          std::string desc = std::regex_replace(desc, regex(average_pattern), \"\");
+
           std::string r = std::regex_replace(std::string(\"abc\"), std::regex(\"b\"), \" \");
           return r == \"a c\" ? 0 : -1;
       }"
```

At the very least, this provides a short and concise repro for testing other fixes. Note that I only tried VS2015. An extra embellishment might be to output a message when boost is used since it's good to know something weird is still going on with regex.

The ordered list of better fixes would be for MS to make their regex more robust to this, to figure out how to increase the _REGEX_MAX_STACK_COUNT  so that it actually is used and to ensure that the new value is good for all types of reasonable expected inputs (100 channel images, maybe 1000, what about a crazy million or billion?), to change the regex in tiffinput to be "nicer", and finally to just give up and use my boost patch. I gave up and went with boost, with the hope that MS will eventually fix this in a few years...

**Comment by ThiagoIze** [comment]
The previous diff doesn't actually compile. This does and also shows how we can more nicely avoid some of the escaping of quotation marks:

```diff
diff --git a/src/cmake/compiler.cmake b/src/cmake/compiler.cmake
index 37a87776..7aabe7a9 100644
--- a/src/cmake/compiler.cmake
+++ b/src/cmake/compiler.cmake
@@ -264,16 +264,37 @@ include (CheckCXXSourceRuns)
 # Find out if it's safe for us to use std::regex or if we need boost.regex
 cmake_push_check_state ()
 set (CMAKE_REQUIRED_DEFINITIONS ${CSTD_FLAGS})
-check_cxx_source_runs("
+
+# Even though we use bracket arguments, we still need to escape the \ characters when passing to check_cxx_source_runs
+check_cxx_source_runs([=[
+//#define _REGEX_MAX_STACK_COUNT 100000
       #include <regex>
       int main() {
-          std::string r = std::regex_replace(std::string(\"abc\"), std::regex(\"b\"), \" \");
-          return r == \"a c\" ? 0 : -1;
-      }"
-      USE_STD_REGEX)
+ 
+          try{
+              // This throws an exception on windows, at least with VS2015, when using std_regex
+              std::string desc="oiio:SHA-1=879F1A17ACE62D0B3D6CCE09B6F2BE3ADD6A48B2 oiio:AverageColor=0.532385,0.49592,0.494299,0,1.16619,0.946394,0.784551,0,0.651571,0.649502,0.65518,0,0.489643,0.471951,0.471693,0,0.599104,0.547551,0.488581,0,0.796633,0.890403,0.84728,0,0.857453,0.874942,0.906667,0,1.26352,1.24764,1.2966,0,1.86804,1.62555,1.16578,0,1.88414,1.9104,2.07001,0";
+              static const char* fp_number_pattern
+                  = "([+-]?((?:(?:[[:digit:]]*\\\\.)?[[:digit:]]+(?:[eE][+-]?[[:digit:]]+)?)))";
+              const std::string average_pattern
+                  = std::string("oiio:AverageColor=(\\\\[?") + fp_number_pattern
+                  + ",?)+\\\\]?[ ]*";
+              desc = std::regex_replace(desc, std::regex(average_pattern), "");
+          }
+          catch(const std::exception& e)
+          {
+             return -1;
+          }
+
+          std::string r = std::regex_replace(std::string("abc"), std::regex("b"), " ");
+          return r == "a c" ? 0 : -1;
+      }
+]=] USE_STD_REGEX)
 if (USE_STD_REGEX)
+    message (STATUS "USE_STD_REGEX")
     add_definitions (-DUSE_STD_REGEX)
 else ()
+    message (STATUS "USE_BOOST_REGEX")
     add_definitions (-DUSE_BOOST_REGEX)
 endif ()
 cmake_pop_check_state ()
```
Now on windows this will use boost unless you uncomment the `#define _REGEX_MAX_STACK_COUNT`, in which case it passes and we use std::regex. So that's good to see that the #define solution does in fact work on its own. However, as before, I don't know why it doesn't work within OIIO itself. I can only assume that the regex header is somehow being called before my #define somehow?

**Comment by ThiagoIze** [comment]
I went with the brute force approach and defined _REGEX_MAX_STACK_COUNT everywhere in OIIO:
```diff
diff --git a/src/cmake/compiler.cmake b/src/cmake/compiler.cmake
index 37a87776..7ceb111b 100644
--- a/src/cmake/compiler.cmake
+++ b/src/cmake/compiler.cmake
@@ -258,22 +258,46 @@ endif ()
 if (NOT VERBOSE)
     set (CMAKE_REQUIRED_QUIET 1)
 endif ()
+
+add_definitions ("-D_REGEX_MAX_STACK_COUNT=0")
+
 include (CMakePushCheckState)
```
and this lets oiiotool work. However, for my application to work (it statically links OIIO and also includes regex for its own use), I also need to define this for all files otherwise I'll get the same error in my OIIO calls. I guess the MS regex header is initializing some sort of global state the first time it's run. Now that I know this trick, that's ok for me, but I suspect that other Windows users that also encounter this will not know they need to do this in their application in order to avoid this bug in OIIO. For this reason I think the best solution might be what I did above and just switch over to boost.

**Comment by lgritz** [comment]
Is perhaps the simplest fix just to change compiler.cmake so that it uses Boost.regex on Windows, until their implementation seems more solid (as we do on gcc when too old to have a fully working std::regex)?

**Comment by ThiagoIze** [comment]
Sure, that works as well, though I'd then suggest adding the above repro to the testsuite.

**Comment by ThiagoIze** [comment]
However, regardless of which you choose, I do suggest switching to the bracket arguments as that would make the preexisting code cleaner since you wouldn't need to escape those quotation marks.

**Comment by lgritz** [comment]
Yeah, definitely. Do you think that your build-time test is sufficient to test for this condition, or should we force boost.regex on Windows just in case? I don't think there is a downside to that, except for the purely aesthetic preference of std over boost whenever possible.

**Comment by meshula** [comment]
It isn't purely aesthetic... Many of us are scouring boost from our builds, and look forward to a brave boost free future. Given the relative simplicity of average_pattern, I want to suggest that a non-regex solution would solve for the out of memory exception, and no new boost dependency simultaneously.

**Comment by ThiagoIze** [comment]
The test above should reliably catch the problem until Microsoft's regex is changed so that it either doesn't use so much memory or they increase the default stack size. Once that happens it's of course possible this particular test will pass but perhaps a similar image with more channels will still fail. My guess is that if MS goes through the effort of changing their code so this passes, they probably will do a good enough job that all reasonable images passed to OIIO will work.

And yes, I too would love for the boost dependency to go away, so this is an unfortunate step back. Fixing the code to not use regex is an option, but that's going to take extra dev time, not to mention it's possible other regex usage in OIIO might also exceed the stack size for certain images.

**Comment by lgritz** [comment]
Proposed fix in #2173 -- I just replaced all uses of regex_replace, @meshula was correct, there was no current use that couldn't be easily substituted with either Strutil::replace or one more short utility function I added as part of this patch.

**Comment by lgritz** [comment]
Closing, this was fixed by #2173.
